### ISSUE TITLE
understanding of maxpoolsize

### ISSUE THREAD

**Issue by yiwong2001** [swaldman/c3p0#93]
Hi there, 

I want to have better understanding of maxpoolSize configuration.

As you saw below, I have 1 master node and 5 slave nodes. Each of nodes can handle max to 1000 connections. 

<bean id="widgetDataSource" class="com.mchange.v2.c3p0.ComboPooledDataSource">  
<property name="driverClass" value="com.mysql.jdbc.ReplicationDriver"/>  
<property name="jdbcUrl" value="jdbc:mysql://master-private:3306,slave1-private:3306,slave2-private:3306,slave3-private:3306,slave4-private:3306,slave5-private:3306/db?autoReconnect=true&amp;failOverReadOnly=false&amp;roundRobinLoadBalance=true&amp;characterEncoding=utf8"/>
 <property name="maxPoolSize" value="**1000**"/>

 I wonder what is value I should set in maxpoolSize. @swaldman, you mentioned before that there would be a maximum of 1000 Connections on the client (I set to 1k now), so total to all servers would be max 1000. If I increase maxpoolSize to 4000, will Master and slave nodes all have 4000/6 each?  I want to set correct number for better performance but do not want to break master node since it can only handle 1000 connections :)  

Can you give me some suggestions?

 Thanks!

**Comment by swaldman** [comment]
1000 is a really big value for `maxPoolSize`. Be sure to set a high `numHelperThreads`. Your performance will probably still suffer from Thread contention.

I don't know what you mean in this context by master and slave nodes. Do you have a replicated database, for each copy of which you keep a separate `DataSource`? I don't understand your topology, so I'm not sure I can answer your questions.

**Comment by yiwong2001** [comment]
Hi swaldman, 

thanks for quick response.

Yes, we are using amazon RDs mysql databases. They are dedicated memory optimized databases. For each of db server, it can handle 1200 max connections. 

We run 1 master database for writing data, and 5 read replicas to sharing read requests. I only configure one database in our application.

**Comment by yiwong2001** [comment]
In addition, what is correct value should set on numHelperThreads based on connections?

**Comment by swaldman** [comment]
So, I think you'll end up with up to 1000 Connections total, with reads quietly directed to your replicas and writes going to your master. You'll want a much bigger `numHelperThreads`, something like 30 instead of 3. (Ideally you'll monitor for a back log in the thread pool via JMX to make sure it's keeping up.)

This is a very large scale for a single DataSource, and Thread contention may well become an issue. To prevent that, you might define multiple DataSources and have clients round-robin or randomly choose between them. (Eventually, I hope to add a facade for multiple pools behind a single `DataSource` facade, but I have not done that yet, so you'd have to implement that logic yourself.) 

Before you go to the trouble of that, give a hard thought to your pool size. Even if your DBMS can handle 1000 simultaneous Connections, you may want to choose a much smaller `maxPoolSize`. See [this](https://github.com/brettwooldridge/HikariCP/wiki/About-Pool-Sizing), by the author of HikariCP. Everything is empirical. I don't know the details of your replicated AWS setup. But you might try with a much more modest `maxPoolSize` (50 - 100), and compare your application's performance to when you have a very large setting. You may find it does not suffer.

**Comment by yiwong2001** [comment]
Thanks @swaldman .. will 30 of numHelperThreads generate 300 threads? is it fine for one single web server?

**Comment by swaldman** [comment]
No, numHelperThreads of 30 will create a thread pool with 30 threads. That's all. There's no obvious relationship between `numHelperThreads` and `acquire increment`, but given so large a pool you might want an `acquireIncrement` a bit larger than its default of 3. Setting `maxPoolSize` and `minPoolSize` the same effectively gives you a fixed-size pool, rather than one that grows and potentially shrinks (depending on other settings) in response to load.
